pkg/scheduler/internal/queue/scheduling_queue_test.go

```diff
@@ -862,105 +862,188 @@ func TestPriorityQueue_Pop(t *testing.T) {
-	// add highPriorityPodInfo to activeQ.
-	q.Update(logger, nil, highPriorityPodInfo.Pod)
-	if _, exists, _ := q.activeQ.Get(newQueuedPodInfoForLookup(highPriorityPodInfo.Pod)); !exists {
-		t.Errorf("Expected %v to be added to activeQ.", highPriorityPodInfo.Pod.Name)
-	}
-	if len(q.nominator.nominatedPods) != 0 {
-		t.Errorf("Expected nomindatePods to be empty: %v", q.nominator)
-	}
-	// Update highPriorityPodInfo and add a nominatedNodeName to it.
-	q.Update(logger, highPriorityPodInfo.Pod, highPriNominatedPodInfo.Pod)
-	if q.activeQ.Len() != 1 {
-		t.Error("Expected only one item in activeQ.")
-	}
-	if len(q.nominator.nominatedPods) != 1 {
-		t.Errorf("Expected one item in nomindatePods map: %v", q.nominator)
-	}
-	// Updating an unschedulable pod which is not in any of the two queues, should
-	// add the pod to activeQ.
-	q.Update(logger, unschedulablePodInfo.Pod, unschedulablePodInfo.Pod)
-	if _, exists, _ := q.activeQ.Get(newQueuedPodInfoForLookup(unschedulablePodInfo.Pod)); !exists {
-		t.Errorf("Expected %v to be added to activeQ.", unschedulablePodInfo.Pod.Name)
-	}
-	// Updating a pod that is already in activeQ, should not change it.
-	q.Update(logger, unschedulablePodInfo.Pod, unschedulablePodInfo.Pod)
-	if len(q.unschedulablePods.podInfoMap) != 0 {
-		t.Error("Expected unschedulablePods to be empty.")
-	}
-	if _, exists, _ := q.activeQ.Get(newQueuedPodInfoForLookup(unschedulablePodInfo.Pod)); !exists {
-		t.Errorf("Expected: %v to be added to activeQ.", unschedulablePodInfo.Pod.Name)
-	}
-	if p, err := q.Pop(logger); err != nil || p.Pod != highPriNominatedPodInfo.Pod {
-		t.Errorf("Expected: %v after Pop, but got: %v", highPriorityPodInfo.Pod.Name, p.Pod.Name)
-	}
 
-	// Updating a pod that is in backoff queue and it is still backing off
-	// pod will not be moved to active queue, and it will be updated in backoff queue
-	podInfo := q.newQueuedPodInfo(medPriorityPodInfo.Pod)
-	if err := q.podBackoffQ.Add(podInfo); err != nil {
-		t.Errorf("adding pod to backoff queue error: %v", err)
-	}
-	q.Update(logger, podInfo.Pod, podInfo.Pod)
-	rawPodInfo, err := q.podBackoffQ.Pop()
-	podGotFromBackoffQ := rawPodInfo.(*framework.QueuedPodInfo).Pod
-	if err != nil || podGotFromBackoffQ != medPriorityPodInfo.Pod {
-		t.Errorf("Expected: %v after Pop, but got: %v", medPriorityPodInfo.Pod.Name, podGotFromBackoffQ.Name)
+	tests := []struct {
+		name      string
+		expectedQ string
+		// addedToNominatedPods is whether a Pod from the test case should be registered as a nominated Pod in the nominator.
+		addedToNominatedPods bool
+		// prepareFunc is the function called before the test calls Update(). After prepareFunc,
+		// each test will call Update() with oldPod and newPod returned from it.
+		prepareFunc func(q *PriorityQueue) (oldPod, newPod *v1.Pod)
+		// validateFunc is the function called after the test calls Update(), to validate the pod is updated as expected.
+		validateFunc func(pInfo *framework.QueuedPodInfo) bool
```

Comment:
Is it necessary to be a function? I mean, can we change it to `wantPod: *framework.QueuePodInfo`?

Reply:
I think we can, I added `wantUpdatedPod` to the return value of `prepareFunc`, as we'll use the `updatedPod` for validate.